src/main/java/com/instaclick/pentaho/plugin/amqp/AMQPPlugin.java

```diff
@@ -100,6 +162,14 @@ public boolean processRow(StepMetaInterface smi, StepDataInterface sdi) throws K
 
                 setOutputDone();
 
+                //now output rowset closed, so we can wait for our confirmation steps
+                if ( data.activeConfirmation )  {
+                    while (!isStopped() && data.watchedConfirmStep.size() > 0) {    
+                        Thread.sleep(500);

```

Comment:
Got it,
Not sure why we need the wait then,

If it is non transactional each row will be ack/nack inside the RowListener,
if transactional is enabled it will be collected by the RowListener and ack/nack as part of the flush.

We don`t even need to keep track of witch step finished or not if its transactional we only check if the transformation succeeded, which is already done by the TransListener.